refactor: move response_content into backend code

nejch opened the pull request:
Needed in order to later add other backends that handle streaming responses differently. Just a first step, later maybe we could get rid of `response_content` completely from individual methods and just handle this in the request code :thinking:

nejch commented on gitlab/_backends/requests_backend.py:
```diff
@@ -9,6 +9,11 @@
 from . import protocol
 
 
+class _StdoutStream:
+    def __call__(self, chunk: Any) -> None:
+        print(chunk)
```
TBH I don't know why we have this. Just doing `action = print` and then `action(chunk)` would do in the function itself. Maybe the author had bigger plans for this.

lmilbaum replied:
It worth adding it as a comment in the code so that contributors could peek it up and change it

nejch replied:
Thanks, I'll open a follow-up issue to track this. Maybe that's a bit cleaner than having comments in the code.

nejch replied:
Opened https://github.com/python-gitlab/python-gitlab/issues/2496

lmilbaum commented on gitlab/_backends/requests_backend.py:
```diff
@@ -117,3 +122,26 @@ def http_request(
         )
         return RequestsResponse(response=response)
+
+    def response_content(
```
The prefix "response_" is redundant

nejch replied:
Keep in mind this method is in `RequestsBackend`, not in `RequestsResponse` as the GitHub diff suggests. Did you still mean this prefix is redundant in the backend class?

lmilbaum replied:
Thanks for the clarification. I totally missed it.  In that case, what is the connection between the `response_content` function and the class `RequestsBackend`?

nejch replied:
The implementation is backend-specific, so it can't live in utils anymore if we want to add other backends. For example requests has `iter_content`, whereas httpx has `iter_bytes` and `iter_text`.

Putting it into the backend class was an easy minimal copy/paste, eventually it would probably make sense to remove this method and handle it in the response itself I assume as I say above:

> Needed in order to later add other backends that handle streaming responses differently. Just a first step, later maybe we could get rid of response_content completely from individual methods and just handle this in the request code :thinking:  

Hope that makes sense:) I can push a branch with an httpx backend to show the different implemenations if needed.

lmilbaum replied:
Thanks again for the clarification. I am totally fine with going forward with this approach as long as we agree that it might be refactored in the future.
Only thing I would add the staticmethod decorator which will eliminate the need of instantiating the class.

nejch replied:
I agree with that, my thinking is as I said,`eventually it would probably make sense to remove this method and handle it in the response itself`, but maybe there are other ways :)

Ah good idea, will add that, completely missed that while copy pasting.

nejch replied:
Done, thanks @lmilbaum!

nejch commented:
@JohnVillalovos I've rebased and reworked this a bit based on previous comments. Would you be able to have a look at this one?

IMO it's also a bit cleaner like this to have the response logic be part of the client/backend than a separate util function since we've split out the client code (and more recently the backend code) from the monolith `__init__` module.

JohnVillalovos reviewed (approved):
Thanks @nejch 

Overall looks good to me. Only a couple of nits on my end. But I would be okay with merging as is. So approved by me.

nejch commented on gitlab/_backends/protocol.py:
```diff
@@ -17,6 +17,17 @@ def __init__(self, response: requests.Response) -> None: ...
 class Backend(Protocol):
+    @staticmethod
+    @abc.abstractmethod
+    def response_content(
```
I believe this would be a breaking change as we still expose it in utils as a non-private function. In that case I'd maybe drop the wrapper entirely and not just deprecate it, as then we don't need to worry about maintaining the signature twice (https://github.com/python-gitlab/python-gitlab/pull/2488#discussion_r1633344488).

Might be worth grouping breaking changes and wait to merge them all together though for a major release.

JohnVillalovos replied:
Well I didn't think it would be a breaking change if this is the new function and then the wrapper (with the deprecation) would call this one with keyword args.

nejch commented on gitlab/utils.py:
```diff
@@ -50,26 +43,14 @@ def format(self, record: logging.LogRecord) -> str:
-    chunk_size: int,
-    *,
-    iterator: bool,
+    *args: Any, **kwargs: Any
```
My initial idea was to just leave a simple wrapper for backward compatibility but maybe we can just drop it (see https://github.com/python-gitlab/python-gitlab/pull/2488#discussion_r1634514002).

JohnVillalovos replied:
My thought was that this one would maintain the same exact signature as before, and then call the new function with keyword arguments.

But! It isn't that important and perfectly fine with me if you want to merge this in as is 👍

nejch replied:
Ahh I misunderstood your comment I think, I thought you wanted to match the signatures. I'll add the changes!
